Fixes #12794 - Defined logic when properties should be sent to server

ShimShtein opened the pull request:
Redefined the logic when the parameters should be sent to server:
If hostgroup is not set - always send the parameters.
On hostgroup change - remove inherited parameters from the request, so the values from the hostgroup will take place.
On save - send the values according to the state of the "inherit" button.

The new tests of this PR are working if I use selenium driver, but when using poltergeist, I see that the test is filling the form, stays stuck at our progress message and never shows the `/hosts/:id` page.

It would be great if the code will be reviewed, in the meantime I will keep on working on forcing poltergeist to work. Right now my main suspect is animation, I am thinking about using code from [here](https://gist.github.com/andyjbas/9962218) as my next step.

ShimShtein committed:
Fixes #12794 - Defined logic when properties should be sent to server

If hostgroup is not set - always send the parameters.
On hostgroup change - remove inherited parameters from the request, so the values from the hostgroup will take place.
On save - send the values according to the state of the "inherit" button.

domcleal commented on test/integration/host_test.rb:
```diff
@@ -69,6 +69,12 @@ def select2(value, attrs)
     end
   end
 
+  def wait_for_ajax
```
This might be best put in test_helper.rb.

domcleal commented on test/integration/host_test.rb:
```diff
@@ -290,6 +294,88 @@ def select2(value, attrs)
+      skip #problem with poltergeist, works on selenium
+      env = FactoryGirl.create(:environment)
+      os = FactoryGirl.create(:ubuntu14_10, :with_associations)
+      Nic::Managed.any_instance.stubs(:dns_conflict_detected?).returns(true)
```
`disable_orchestration`?

domcleal commented on app/helpers/hosts_and_hostgroups_helper.rb:
```diff
@@ -115,4 +115,13 @@ def explicit_value?(field)
+    # if the host has no hostgroup
+    return true unless @host && @host.hostgroup
+    # when editing a host, the values are specified explicitly
+    return true if params[:action] == 'edit'
```
Probably also `clone`?

domcleal commented:
1) Could you include the info about the different logic (first few lines of the PR description) in the commit message for future reference?

ShimShtein commented:
Addressed all the issues, tests fixed and rebased to rails 4

ShimShtein committed:
Fixes #12793 - reverted field filtration to enabled fields only

ShimShtein commented:
Added a commit that replaces #2972.
Reverted the filtration logic to enabled only fields, and changed `reset_explicit_values` function to disable fields instead of setting `explicit` attribute.
Plus added a test for [#12793](http://projects.theforeman.org/issues/12793) that now is passing.

domcleal commented on app/helpers/hosts_and_hostgroups_helper.rb:
```diff
@@ -61,7 +61,7 @@ def puppet_ca(f, can_override, override)
                :disable_button => can_override ? _(INHERIT_TEXT) : nil,
                :disable_button_enabled => override && !explicit_value?(:puppet_ca_proxy_id),
-               :user_set => params[:host] && params[:host][:puppet_ca_proxy_id]
+               :user_set => user_set?(:puppet_ca_proxy_id)
```
Is `:user_set` now obsolete, as the JS no longer checks data-explicit?

domcleal replied:
Ah no, we're still using it in set_inherited_value.

domcleal commented:
Merged as d94f4d3, 77b36c9, thanks @ShimShtein.
